### PR TITLE
topology: get nhlt first from bios if exists

### PR DESCRIPTION
**Pull request by juimonen**
fixup! topology: set dmic vendor blob from nhlt

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

### CODE REVIEW

**Review by ranj063 — APPROVED** [review]

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -2002,6 +2011,10 @@ static int sof_manifest(struct snd_soc_component *scomp, int index,
 		}
 	}
 
+	/* get nhlt from bios if exists and no override set */
+	if (!sof_use_tplg_nhlt)
+		sdev->nhlt = intel_nhlt_init(sdev->dev);
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2016]: this makes me cringe. The intel NHTL stuff should be Intel-specific, I don't think it's wise to have this code in the SOF core.

**Review by plbossart — COMMENTED** [review]
this needs to be rethought @juimonen @ranj063, see below.

File: sound/soc/sof/topology.c (hunk 1 above)
Comment (ranj063) [reply sound/soc/sof/topology.c:2016]: @plbossart makes sense. We will move it to the intel-specific code where we do the check for DMIC's

**Comment by bardliao** [comment]
Note that we need to set `options snd_sof sof_use_tplg_nhlt=1` on our CI devices to test with this PR.

**Comment by bardliao** [comment]
Maybe it is better to fetch nhlt from topology by default?

**Comment by ranj063** [comment]
> Maybe it is better to fetch nhlt from topology by default?

@bardliao no we want the NHLT from BIOS by default and only override it if we want the tplg NHLT with the kernel module param.

**Comment by XiaoyunWu6666** [comment]
@ranj063 @bardliao Does it mean that we need to use kernel module param to get tplg NGLT for all IPC4 models?

**Comment by XiaoyunWu6666** [comment]
hi @ranj063  do you suggest use this kernel module param for all IPC4 model that use nocodec topology?
I just found all IPC4 model that use nocodec topology failed to prepare widget copier.SSP when aplay with this commit.
And with the kernel param ```options snd_sof sof_use_tplg_nhlt=1``` , they are good

**Comment by ranj063** [comment]
> hi @ranj063 do you suggest use this kernel module param for all IPC4 model that use nocodec topology? I just found all IPC4 model that use nocodec topology failed to prepare widget copier.SSP when aplay with this commit. And with the kernel param `options snd_sof sof_use_tplg_nhlt=1` , they are good

@XiaoyunWu6666 yes, for all nocodec tplg's. For all HDA/SDW devices, we should not set this param.

**Comment by bardliao** [comment]
@ranj063 Can we check if BIOS provides nhlt probably? Or at least check if the nhlt  exist.

**Comment by ranj063** [comment]
> @ranj063 Can we check if BIOS provides nhlt probably? Or at least check if the nhlt exist.

@bardliao if there's no NHLT in BIOS intel_nhlt_init() would return NULL and we would fallback to the one in topology.

**Comment by bardliao** [comment]
> > @ranj063 Can we check if BIOS provides nhlt probably? Or at least check if the nhlt exist.
> 
> @bardliao if there's no NHLT in BIOS intel_nhlt_init() would return NULL and we would fallback to the one in topology.

Got you, thanks :)

**Pull request merged**